### PR TITLE
Enrich sed with more functions, address support, hold/pattern space and more

### PR DESCRIPTION
**Pull request by rtobar**
**History**: This work is based on my original PR under https://github.com/SerenityOS/serenity/pull/10657 where I introduced the `sed` utility. That PR however went stale and another one introduced `sed` has now been created under https://github.com/SerenityOS/serenity/pull/16584. This PR sits on top of the latter, but is based on the contents of the former.

This PR enriches our `sed` implementation by adding support for:
 * Parsing *almost* all commands and their arguments.
 * Executing *most* commands, not just the `s` command.
   * This also implies implementing the pattern and hold spaces.
 * Parsing and obeying the optional/required address ranges for each command.
 * `unveil`-ing all required input/output files.
 * Accepting multiple commands, either via command line arguments, or via script files.
 * Improved error reporting.

On top of this, a second commit adds a humble set of unit tests for `sed`. To implement these tests I've written a small `Process` class based mostly on similar code found in `test262` that simplifies executing a command with a given command line and stdin contents, and returns its stdout .and stderr. I recognise this particular `Process` class is very limited, but I think it would eventually be good to improve it (use EventLoop, better API, etc) and move it under `LibCore` (and then reuse it here, in `test262`, and probably more places as it would be a handy little tool). A fight for another day though.

This PR also contains a fix for a corner case with `BufferedFile` against a file with no newlines, which returned no contents when iterating with `can_read_line()`, which in turn affected programs line sed, grep, sort, etc.

An example of what you can do now:
```
$ echo hellloooooooooo | sed -n 'h; x; s/./*/gp; x; h; p; x; s/./*/gp'
***************
hellloooooooooo
***************
$
```

### CODE REVIEW

File: Userland/Utilities/sed.cpp [hunk 1]
```diff
@@ -14,138 +20,866 @@
...
+
+    bool matches([[maybe_unused]] StringView pattern_space, size_t line_number, bool is_last_line) const
+    {
+        switch (m_address_type) {
+        case AddressType::Line:
+            return line_number == m_line_number;
+        case AddressType::LastLine:
+            return is_last_line;
+        default:
+            Core::File::standard_error()->write(DeprecatedString::formatted("Addressing type not implemented: {}", int(m_address_type)));
```
Comment (alimpfard) [inline Userland/Utilities/sed.cpp:143]:
```suggestion
            warnln("sed: Addressing type not implemented: {}", to_underlying(m_address_type));
```

Comment (rtobar) [reply Userland/Utilities/sed.cpp]: Thanks @alimpfard, I've changed all stderr calls to use `warnln`, and most stdout calls to use `outln` (I'm still passing the stdout `File` object around in some places explicitly as sometimes outputs goes into a file instead).

**Comment by rtobar** [comment]
I've finally been able to dedicate some time to this PR again, rebasing it on the latest master and fixing some pending porting bits (`Stream::File` is just `File` now, etc.). I've also included a fix for a corner case in `BufferedHelper::can_read_line`, which wasn't working properly on files without newlines, reporting there were no lines to read. This not only affected sed, but other programs that used the `can_read_line` + `read_line` combo to progress through lines of their input files.

**Review by konradekk — COMMENTED** [review]
iʼd go for more explicit names in the code for the commands; i understand that ‘aʼ or ‘bʼ are perfectly valid for reference coherence and users versed with the usage, but iʼd (personally) try to keep those only to cli and parser naming functionalities just after this point (also producing enums for them?).

yet: (a) thatʼs my personal taste, (b) this can wait, (c) itʼs not really that important; so… resolve according to your own discern. 🙇🏻‍♂️

**Comment by rtobar** [comment]
@konradekk thanks for the review. I agree that more descriptive names are definitely better. I went and had a look to change them.... only to find they are not used! Given that, I opted for *not* adding them, as they would actually add a layer of indirection between reading the code and understanding which function (i.e., which character in the sed stream) a particular line is referring to. Happy to hear other thoughts anyway (e.g., if you think adding the names back, and actually using them).

**Comment by konradekk** [comment]
@rtobar: i think i have nothing to add! 😅

your approach is reasonable—my only thought is that if you read the code, you might want to know what are looking at without going into details. you might want to know, by reading a function name, that it should implement a substitution feature, yet itʼs also sensible to say, that what you see at given moment is an implementation of ‘sʼ function requested in the arguments.

i could encourage you to use comments, but on the other hand i was conditioned to believe that one should use meaningful names instead (yet another adages: code is written way less times than itʼs read, and: comments tend to be ignored when when changing the code theyʼre related to). 


——as you asked for ‘my other thoughts,ʼ i took liberty to write them down here! 😂 (please skip this section if not interested!)——

> my approach is to indicate points i believe/heard might be problematic now or in the future. it is you to decide whether the code needs change or not. 🙂 you definitely need not to convince me to anything as iʼm not a person to be convinced here! 😅 (that burden is on maintainers!)
>
> nowadays programmers are encouraged to work in iterations to make progress, continuously integrate their work with others so all collaborators share the same… knowledge. the other purpose of me reviewing is to ask the questions so i can learn something myself. but it is not only about myself—i believe that knowledge gain can be a share of all parties: the reviewer(s), author(s), and thus of the group as a whole. as you might know, thereʼs a small problem with this: knowledge needs to be used—it canʼt be just gathered. so review outcome must be a code that is acceptable, that makes a progress. and of course learning (acquiring knowledge) is a pain, so it must be digestible! 😄 that means that author needs to clearly set oneself what is the iteration one wants to achieve—as it was pointed out to me recently, thatʼs also a responsibility of the reviewer.
>
> when reading iʼm trying understand the assumptions the code was written under. yet i like have them stated by the author explicitly as many mistakes or miscomprehensions (probably also between people in social interactions) happen because of those being misaligned. this reveals itself in my questions like: have you thought about this? have you took that into account? realising that in most cases help to clear the expectations and understand the subject better. and by extension: share the knowledge, understand people better. (well, i just understood that iʼve got pretty much a philosophical approach to that! not the best thing in the software development, i guess! 😂)


i think the bottom line (tl;dr): use your best judgment. i am just a poor wayfaring stranger! 🙂

(and probably: and donʼt ask me about my thoughts! 😂)

**Review by ADKaster — APPROVED** [review]
I agree with the comment in the description about the process utility. We've got implementations in LibCore, test262-runner, run-tests.cpp that all likely want to do similar things: spawn a process, and either interact with it through pipes/tmp files, or read its output during/after execution. Maybe worth an issue :D

Regarding the flags, I agree it's a bit hard to follow with the current class names. However, looking at the man page for GNU or BSD sed, the options themselves are very obscure and interconnected, and hard to describe with a pithy name.

Speaking of the man pages, a follow-up to update the man page is probably worthwhile :)

**Comment by rtobar** [comment]
Thanks @konradekk again for the review and further discussion. I salute your dedication to code reviews by the way, both in terms of quantity and quality; the former I learned by having a quick at your profile, while from your last comment it's very clear you put a lot of thought into the process. Just to touch on some of your final points: yes, I'm trying to stay away from comments by having descriptive names (but it's a difficult art to master!), and I agree with having incremental progress. This version of `sed` itself is not finished, and there still are many places for improvement, which I hope will be filled in time. I know this isn't the place to have long discussions, so I'll close with saying that I also found interesting your thoughts about the role that authors and reviewers have, and how they interplay.

And thanks to @ADKaster for merging this in. I'll raise an issue for a centralised `Process` class, and will also work on a manpage for `sed` (it hadn't even crossed my mind until now!)
